leapp/utils/clicmd.py

```diff
@@ -89,7 +89,15 @@ def execute(self, version):
         else:
             s = parser
         self.apply_parser(s, parser=parser)
-        args = parser.parse_args()
+        # XXX FIXME a clumsy hack to make leapp output help information for the
+        # exact subcommand that has been called.
+        # The original problem was with parser.parse_args() returning parser for
+        # topmost leapp command and not the subcommand.
+        # Any ideas and proposals on a more elegant solution are welcome.
+        args, leftover = parser.parse_known_args()
+        if leftover:
+            cmdname = args.prog.split()[-1]
+            self._sub_commands[cmdname].parser.error(leftover)
```

Comment:
You should handle the case that cmdname is wrong otherwise you get a KeyError here.

Reply:
Actually (I was surprised as well) in case of `leapp` or `leapp nosuchsubcommand` the exception will be raised during parse_known_args() execution. So the only time lines L98-L100 are reached is when a proper leapp subcommand has been called.

Reply:
Worth a comment then, for future readers wondering why it's not handled

Reply:
For the peace of mind gave it another thorough test:
- leapp no arguments
```
[root@ivasilev-propersubcommandhelp ~]# leapp
usage: leapp [-h] [--version] [--verbose] [--debug]  ...

Optional arguments:
  -h, --help  show this help message and exit
  --version   show program's version number and exit
  --verbose   Enable verbose logging
  --debug     Enable debug mode

Main commands:
  
    upgrade   Upgrades the current system to the next available major version.
    list-runs
              List previous Leapp upgrade executions
    preupgrade
              Generate preupgrade report
error: too few arguments
```
- leapp badsubcommand
```
[root@ivasilev-propersubcommandhelp ~]# leapp nosuchcommand
usage: leapp [-h] [--version] [--verbose] [--debug]  ...

Optional arguments:
  -h, --help  show this help message and exit
  --version   show program's version number and exit
  --verbose   Enable verbose logging
  --debug     Enable debug mode

Main commands:
  
    upgrade   Upgrades the current system to the next available major version.
    list-runs
              List previous Leapp upgrade executions
    preupgrade
              Generate preupgrade report
error: argument : invalid choice: 'nosuchcommand' (choose from 'upgrade', 'list-runs', 'preupgrade')
```
- leapp upgrade --nosuchflag
```
[root@ivasilev-propersubcommandhelp ~]# leapp upgrade --nosuchflag
usage: leapp upgrade [-h] [--whitelist-experimental ActorName] [--reboot]
                     [--resume] [--verbose] [--debug]

Description:

Upgrades the current system to the next available major version.

Optional arguments:
  -h, --help            show this help message and exit
  --whitelist-experimental ActorName
                        Enables experimental actors
  --reboot              Automatically performs reboot when requested.
  --resume              Continue the last execution after it was stopped (e.g.
                        after reboot)
  --verbose             Enable verbose logging
  --debug               Enable debug mode
error: ['--nosuchflag']
```

Reply:
I think that's ok. It's true that in the last case (--nosuchflag) could be better to see proper message like: **No such option exists** or something like that. But I think this is still clear and the solution is elegant :)